pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,157 @@
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = false
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = true
+	}
+}
+
+func (o *APIResourceEnablementConfig) DisableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Delete(resource.Resource)
+		o.GroupVersionDisablers[version].DisabledResources.Insert(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Insert(resource.Resource)
+		o.GroupVersionDisablers[version].DisabledResources.Delete(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) ResourceEnabled(resource unversioned.GroupVersionResource) bool {
+	if resource.Version == AnyVersion {
+		return o.anyVersionOfResourceEnabled(resource)
+	}
+
+	return o.exactResourceEnabled(resource)
+}
+
+func (o *APIResourceEnablementConfig) anyVersionOfResourceEnabled(resource unversioned.GroupVersionResource) bool {
+	for version, _ := range o.GroupVersionDisablers {
+		if version.Group != resource.Group {
+			continue
+		}
+
+		if o.exactResourceEnabled(version.WithResource(resource.Resource)) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (o *APIResourceEnablementConfig) exactResourceEnabled(resource unversioned.GroupVersionResource) bool {
+	versionOverride, versionExists := o.GroupVersionDisablers[resource.GroupVersion()]
+	if !versionExists {
+		return true
+	}
+	if !versionOverride.Enable {
+		return false
+	}
+
+	if versionOverride.DisabledResources.Has(resource.Resource) {
+		return false
+	}
+
+	if len(versionOverride.EnabledResources) > 0 {
+		return versionOverride.EnabledResources.Has(resource.Resource)
+	}
+
+	return true
+}
+
+func (o *APIResourceEnablementConfig) VersionEnabled(version unversioned.GroupVersion) bool {
+	versionOverride, versionExists := o.GroupVersionDisablers[version]
+	if !versionExists {
+		return false
```

Comment:
isn't it odd for `exactResourceEnabled(mygroup/myversion/myresource)` to return `true`, and `VersionEnabled(mygroup/myversion)` to return `false`?